### ISSUE TITLE
CLI Beta Launch

### ISSUE THREAD

**Issue by CMCDragonkai** [MatrixAI/Polykey-CLI#40]
### Specification

This epic focuses on the CLI beta launch targeting November 10 2023.

This follows from the 6th testnet deployment https://github.com/MatrixAI/Polykey/issues/551 and focuses on any UX issues that should be resolved before we launch. Plus any documentation related things and metrics for tracking how the launch went, as well as content we need to write to prepare for it.

We should also get our demo video fully polished as well.

One of the things we need to do is:

1. Add a `audit` or `logs` domain to PK at first to track information about the network, we will need this as part of our metrics.
2. Create a simple dashboard on `testnet.polykey.com` and `mainnet.polykey.com` so we can see what's going on and show everybody how the network is building up.

We are currently working through this list of issues.
https://github.com/MatrixAI/Polykey-CLI/issues/40#issuecomment-1853222990

### Additional context

* https://github.com/MatrixAI/Polykey/issues/551

### Tasks

1. Spec out the experience for the launch
4. Schedule content releases
5. Plan out all UX-related issues

**Comment by CMCDragonkai** [comment]
https://github.com/MatrixAI/Polykey/pull/600 was merged incorrectly, https://github.com/MatrixAI/Polykey/pull/601 is being done by @addievo to fix up the additional options of `privateKey` and `privateKeyPath` and `recoveryCode`. Also they would need to be exposed on `PolykeyAgent.start` too. See review comments in https://github.com/MatrixAI/Polykey/pull/600

**Comment by tegefaulkes** [comment]
I finished up and merged https://github.com/MatrixAI/Polykey/pull/601. Having a look over the CLI now.

**Comment by CMCDragonkai** [comment]
The `PolykeyClient` is going to be create/destroy only, I think it's pretty clear that can work.

**Comment by CMCDragonkai** [comment]
The change to the `GitHubProvider` to using graphql for getting connected identities is now causing occassional failures in the `IdentitiesManager.test.ts`. I've seen it on CI and also locally.

```
 FAIL  tests/identities/IdentitiesManager.test.ts (10.559 s)
  IdentitiesManager
    ✓ IdentitiesManager readiness (154 ms)
    ✓ get, set and unset tokens (with seed=1486144507) (1350 ms)
    ✓ start and stop preserves state (with seed=1486144507) (966 ms)
    ✓ fresh start deletes all state (with seed=1486144507) (922 ms)
    ✓ register and unregister providers (103 ms)
    ✓ using TestProvider (with seed=1486144507) (3075 ms)
    ✕ handleClaimIdentity (with seed=1486144507) (2837 ms)

  ● IdentitiesManager › handleClaimIdentity (with seed=1486144507)

    Property failed after 1 tests
    { seed: 1486144507, path: "0:0:0:0:0:0:0:0:0:0:0:0:0:3:0:1:1:0:3:0:0:0:0:1:2:1:1:1:2:1", endOnFailure: true }
    Counterexample: ["",{"accessToken":"          ","refreshToken":"","accessTokenExpiresIn":1698180425,"refreshTokenExpiresIn":0}]
    Shrunk 29 time(s)
    Got ErrorProviderUnauthenticated: Access token expired

      70 |     ) {
      71 |       if (!providerToken.refreshToken) {
    > 72 |         throw new identitiesErrors.ErrorProviderUnauthenticated(
         |               ^
      73 |           'Access token expired',
      74 |         );
      75 |       }

      at TestProvider.checkToken (src/identities/Provider.ts:72:15)
      at TestProvider.checkToken [as publishClaim] (tests/identities/TestProvider.ts:152:16)
      at src/identities/IdentitiesManager.ts:272:39
      at constructor_.addClaim (src/sigchain/Sigchain.ts:449:20)
      at withF (node_modules/@matrixai/resources/src/utils.ts:24:12)
      at constructor_.handleClaimIdentity (src/identities/IdentitiesManager.ts:261:5)
      at numRuns (tests/identities/IdentitiesManager.test.ts:361:7)
      at AsyncProperty.run (node_modules/fast-check/lib/check/property/AsyncProperty.generic.js:49:28)
      Hint: Enable verbose mode in order to have the list of all failing values encountered during the run
      at buildError (node_modules/fast-check/lib/check/runner/utils/RunDetailsFormatter.js:131:15)
      at asyncThrowIfFailed (node_modules/fast-check/lib/check/runner/utils/RunDetailsFormatter.js:148:11)
```

@amydevs

**Comment by amydevs** [comment]
> The change to the `GitHubProvider` to using graphql for getting connected identities is now causing occassional failures in the `IdentitiesManager.test.ts`. I've seen it on CI and also locally.
> 
> ```
>  FAIL  tests/identities/IdentitiesManager.test.ts (10.559 s)
>   IdentitiesManager
>     ✓ IdentitiesManager readiness (154 ms)
>     ✓ get, set and unset tokens (with seed=1486144507) (1350 ms)
>     ✓ start and stop preserves state (with seed=1486144507) (966 ms)
>     ✓ fresh start deletes all state (with seed=1486144507) (922 ms)
>     ✓ register and unregister providers (103 ms)
>     ✓ using TestProvider (with seed=1486144507) (3075 ms)
>     ✕ handleClaimIdentity (with seed=1486144507) (2837 ms)
> 
>   ● IdentitiesManager › handleClaimIdentity (with seed=1486144507)
> 
>     Property failed after 1 tests
>     { seed: 1486144507, path: "0:0:0:0:0:0:0:0:0:0:0:0:0:3:0:1:1:0:3:0:0:0:0:1:2:1:1:1:2:1", endOnFailure: true }
>     Counterexample: ["",{"accessToken":"          ","refreshToken":"","accessTokenExpiresIn":1698180425,"refreshTokenExpiresIn":0}]
>     Shrunk 29 time(s)
>     Got ErrorProviderUnauthenticated: Access token expired
> 
>       70 |     ) {
>       71 |       if (!providerToken.refreshToken) {
>     > 72 |         throw new identitiesErrors.ErrorProviderUnauthenticated(
>          |               ^
>       73 |           'Access token expired',
>       74 |         );
>       75 |       }
> 
>       at TestProvider.checkToken (src/identities/Provider.ts:72:15)
>       at TestProvider.checkToken [as publishClaim] (tests/identities/TestProvider.ts:152:16)
>       at src/identities/IdentitiesManager.ts:272:39
>       at constructor_.addClaim (src/sigchain/Sigchain.ts:449:20)
>       at withF (node_modules/@matrixai/resources/src/utils.ts:24:12)
>       at constructor_.handleClaimIdentity (src/identities/IdentitiesManager.ts:261:5)
>       at numRuns (tests/identities/IdentitiesManager.test.ts:361:7)
>       at AsyncProperty.run (node_modules/fast-check/lib/check/property/AsyncProperty.generic.js:49:28)
>       Hint: Enable verbose mode in order to have the list of all failing values encountered during the run
>       at buildError (node_modules/fast-check/lib/check/runner/utils/RunDetailsFormatter.js:131:15)
>       at asyncThrowIfFailed (node_modules/fast-check/lib/check/runner/utils/RunDetailsFormatter.js:148:11)
> ```
> 
> @amydevs

this is failing because we are taking we are using the falsiness of the values in the statement rather than matching `== null`:

```
public async checkToken(
    providerToken: ProviderToken,
    identityId?: IdentityId,
  ): Promise<ProviderToken> {
    const now = Math.floor(Date.now() / 1000);
    // this will mean that accessTokenExpiresIn = 0 will be false
    if (
      providerToken.accessTokenExpiresIn &&
      providerToken.accessTokenExpiresIn >= now
    ) {
      // this will mean that refreshToken = '' will throw
      if (providerToken.refreshToken) {
        throw new identitiesErrors.ErrorProviderUnauthenticated(
          'Access token expired',
        );
      }
      // this will mean that refreshTokenExpiresIn = 0 does not throw
      if (
        providerToken.refreshTokenExpiresIn &&
        providerToken.refreshTokenExpiresIn >= now
      ) {
        throw new identitiesErrors.ErrorProviderUnauthenticated(
          'Refresh token expired',
        );
      }
      return await this.refreshToken(providerToken, identityId);
    }
    return providerToken;
  }
```

This would mean that a value of 0 for the `ExpiresIn`mean that they never expire.

Is this behaviour intended? If so I'll keep that as it is.

**Comment by CMCDragonkai** [comment]
I think you found out that 0 means infinity.

**Comment by CMCDragonkai** [comment]
We got about 2 weeks, so let's focus hard on getting stability and bugs fixed.

**Comment by CMCDragonkai** [comment]
A bunch of different issues have been thrown into this epic. We have to go over prioritisation of which is most important and delegate the tasks. Remember we got a tight timeframe here.

**Comment by tegefaulkes** [comment]
Critical issues
1. https://github.com/MatrixAI/Polykey/issues/592
2.  https://github.com/MatrixAI/Polykey/issues/597
3. https://github.com/MatrixAI/js-rpc/issues/47 https://github.com/MatrixAI/Polykey/issues/588 https://github.com/MatrixAI/Polykey/issues/589
4. https://github.com/MatrixAI/Polykey-CLI/issues/44 


needed but functional without
1. https://github.com/MatrixAI/Polykey/issues/598
5. https://github.com/MatrixAI/Polykey/issues/605
1. #17 
6. https://github.com/MatrixAI/Polykey/issues/572
7. https://github.com/MatrixAI/Polykey/issues/593
8. #37
9. https://github.com/MatrixAI/Polykey/issues/594, #38
10. https://github.com/MatrixAI/Polykey/issues/608
11.  #22 
12. #518 - Is this even still a bug? I think we would've seen it during testnet testing 

stretch
1. #36 
3. https://github.com/MatrixAI/Polykey/issues/599
4. #30 
5. https://github.com/MatrixAI/Polykey/issues/237
6. https://github.com/MatrixAI/Polykey-CLI/issues/31
7. #6 
8. https://github.com/MatrixAI/Polykey/issues/591



(wip)

**Comment by CMCDragonkai** [comment]
Stretch 8. should be done to make sure timeouts are working.

> Add a audit or logs domain to PK at first to track information about the network, we will need this as part of our metrics.

No issue for this, but it's really part of https://github.com/MatrixAI/Polykey/issues/599, and we will need that for metrics for the launch, can't launch without some metrics.

**Comment by CMCDragonkai** [comment]
We need to solve that https://github.com/MatrixAI/Polykey/issues/598 to ensure that we don't get hit with a hug of death.

**Comment by CMCDragonkai** [comment]
https://github.com/MatrixAI/Polykey/issues/605 - regardless of how strong the strong NAT - it should work either way.

**Comment by tegefaulkes** [comment]
I'm going to focus on fixing https://github.com/MatrixAI/js-timer/issues/15 really quick. It's causing some minor issues across the code bases.

**Comment by tegefaulkes** [comment]
Not sure how to move forward with https://github.com/MatrixAI/js-timer/issues/15, It's an easy fix assuming my assumptions about how it should work is correct. But `js-timer` has been updated to ESM. So I can't make a new release we can use.

**Comment by CMCDragonkai** [comment]
Timer is fixed.

**Comment by CMCDragonkai** [comment]
For the CLI beta launch to scale, because potentially many people may try PK and thus connect to the testnet, we need to avoid connecting to every seed node in the testnet at the same time. However this means hole punching won't work without a common seed node between source and target. To achieve this we need to revisit the decentralized hole punching problem in https://github.com/MatrixAI/Polykey/issues/365

**Comment by CMCDragonkai** [comment]
Task assignments - for 1/2 days.

@amydevs 

Firstly to fix up the timeouts for RPC, and then apply to both the agent service and the github auth timeout issue.

* https://github.com/MatrixAI/js-rpc/pull/42
* https://github.com/MatrixAI/js-rpc/pull/47
* https://github.com/MatrixAI/Polykey/issues/572
* https://github.com/MatrixAI/Polykey/issues/588
* https://github.com/MatrixAI/Polykey/pull/589
* https://github.com/MatrixAI/js-rpc/issues/19

And then update the PK CLI - test the github auth timeout.

@tegefaulkes 

* https://github.com/MatrixAI/Polykey/issues/592
* https://github.com/MatrixAI/Polykey/issues/597
* https://github.com/MatrixAI/Polykey/issues/598
* https://github.com/MatrixAI/Polykey/issues/593
* https://github.com/MatrixAI/Polykey/issues/605

@okneigres 

* https://github.com/MatrixAI/Polykey/issues/599

@addievo 

* https://github.com/MatrixAI/Polykey-CLI/pull/45
* https://github.com/MatrixAI/Polykey-CLI/issues/44
* https://github.com/MatrixAI/Polykey-CLI/issues/30
* https://github.com/MatrixAI/Polykey/issues/564
* https://github.com/MatrixAI/Polykey-CLI/issues/31

**Comment by amydevs** [comment]
currently Polykey agents domain do not use ctx at all, so server-side handlers will never time out. This will need to be done, and tests will need to be written with timeouts in that domain

**Comment by CMCDragonkai** [comment]
@okneigres first requirement of PKE is just the testnet/mainnet pages. The second requirement of it running the seed cluster is less important, but would be nice to have as it makes it easier to do complex orchestration.

**Comment by CMCDragonkai** [comment]
Almost forgot, but the major thing to solve here is an `audit` domain so that we can record high level events and report that to the dashboard.

@okneigres

I can start the work but @amydevs or @tegefaulkes needs to take over to do that. Could be relevant to #386.

**Comment by CMCDragonkai** [comment]
Solving https://github.com/MatrixAI/Polykey/issues/599 requires switching to using SRV records, and at an intermediate stage we won't have hardcoded trusted seed nodes in our `config.ts` in PK, and we can add back in trusted seed nodes after we have CSR capability https://github.com/MatrixAI/Polykey/issues/154

**Comment by CMCDragonkai** [comment]
@amydevs eta on #50, #53, #49? They seem to be stacked one after another.

**Comment by CMCDragonkai** [comment]
@tegefaulkes it's time to start addressing 

![image](https://github.com/MatrixAI/Polykey-CLI/assets/640797/80825ff4-9330-4ea0-afa3-496b95b8479c)

Just make sure the majority of the memory leak is contained, we can still continue with a minor leak - lower growth rate, and put those as a later priority.

**Comment by amydevs** [comment]
> @amydevs eta on #50, #53, #49? They seem to be stacked one after another.

@CMCDragonkai should be all doable by today

**Comment by CMCDragonkai** [comment]
10th - status update

@amy

* #30 - should be doable by following the idea that all vaults should be shareable within the same gestalt by default - spec out the tasks
* #36 and #52 should be done after starting on the `audit` domain because that's where you're going to be understanding the high level events (that is summarised into metrics).
  - Think of events as independent units of data.
  - Metric are summarisations of all the units
  - Prefer to keep the units around, unless we don't need the data
  - Computing the metrics can be a running-algorithm, so doesn't require always doing it from the beginning
* #22 - requires following up on the remaining tasks, any special formatting outside of dict or table will require case by case analysis
* Main thing is the `audit` domain. Remember to finish off all of those above, except #36 and #52.
* CI failure for docker tests for PK CLI - https://gitlab.com/MatrixAI/open-source/Polykey-CLI/-/jobs/5498527101

@tegefaulkes

* https://github.com/MatrixAI/Polykey/issues/614 - can be looked at while working on the nodes domain
* https://github.com/MatrixAI/Polykey/issues/597 - TBD
* https://github.com/MatrixAI/Polykey/issues/593#issuecomment-1804804909 - review that the order of printing certs is correct from root to leaf, while array order is leaf to root
* https://github.com/MatrixAI/Polykey/issues/598 - requires an empirical test on the testnet, provide results on the current run with the latest fix to the logger - staging is automatically deployed to the testnet - it needs to be checked
* https://github.com/MatrixAI/Polykey/issues/605 - https://github.com/MatrixAI/Polykey/pull/624 - this isn't solved.

@addievo 

* No more discord bot
* Polykey.com/api - what authentication is needed for the github metrics?

**Comment by CMCDragonkai** [comment]
Remember everything is auto deployed on testnet. WE need to do empirical testing on all of the problems now.

**Comment by tegefaulkes** [comment]
Just a note,  while verification logic will handle cert chains in order of leaf -> root. The PEM format should be ordered in root -> leaf. This will need to be checked. @amydevs

**Comment by tegefaulkes** [comment]
I've I have an Ubuntu VM I can access from the tailscale network now. I'm trying to test the linux version of `Polykey-CLI` to test hole punching between networks. 

Running the program results in the following error..

```
XYZ@pttest:~$ ./result
pkg/prelude/bootstrap.js:2255
    return ancestor.dlopen.apply(process, args);
                           ^

Error: /lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /tmp/pkg/c65c5bc459c9452b15357f0d369812780597139e731512beb6903949ed560ba3/@matrixai/db/prebuilds/linux-x64/node.napi.node)
    at process.dlopen (pkg/prelude/bootstrap.js:2255:28)
    at Module._extensions..node (node:internal/modules/cjs/loader:1338:18)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at Module.require (pkg/prelude/bootstrap.js:1851:31)
    at require (node:internal/modules/cjs/helpers:110:18)
    at load (/snapshot/Polykey-CLI/node_modules/@matrixai/db/node_modules/node-gyp-build/index.js:22:10)
    at Object.<anonymous> (/snapshot/Polykey-CLI/node_modules/@matrixai/db/dist/native/rocksdb.js:8:46)
    at Module._compile (node:internal/modules/cjs/loader:1254:14) {
  code: 'ERR_DLOPEN_FAILED'
}

Node.js v18.15.0

```

It seems to be missing a dynamic library. I don't know if this is a problem with the building/bundling process. It's looking for it in `/lib/` so I don't think it would be included in the package.

Is this just something we need to install for `Polykey` to run?

**Comment by CMCDragonkai** [comment]
See https://chat.openai.com/share/be6a1745-59e7-4e76-b5a4-2f7f02f7f3bc

Non-NixOS suck. That's why nobody uses it.

**Comment by CMCDragonkai** [comment]
That is a C++ standard library required by the latest build of rocksdb.

```
Error: /lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /tmp/pkg/c65c5bc459c9452b15357f0d369812780597139e731512beb6903949ed560ba3/@matrixai/db/prebuilds/linux-x64/node.napi.node)
```

Most OS distributions would need to package manage such things, whenever the standard library of C++ is updated. RocksDB is a fast moving project most likely. We don't have resources to deal with people on old machines/VMs.

Do your tests on NixOS preferably.

**Comment by CMCDragonkai** [comment]
@amydevs progress update:

> was half way through the sensitive replacer for errors, it should be done. Just need to confirm edgecases with tests. Other than that, the first 2 commits of STDOUT standardization are mergable, the 3rd commit is a WIP that I wan't to batch up a bunch of changes in, cos they are largely all have the same things that need to be fixed.

> And was still reading through GestaltGraph and Discovery to get a good idea on how to get GestaltID sharing working

**Comment by CMCDragonkai** [comment]
@amydevs will need you to start speccing out the `audit` domain.

**Comment by CMCDragonkai** [comment]
![image](https://github.com/MatrixAI/Polykey-CLI/assets/640797/37a3299e-56a7-4195-a461-7f5559a4fb95)

We will need coordination between 3 areas:

1. The dashboard testnet/mainnet service
2. The polykey.com/api CF worker that has to call the testnet/mainnet service to get the data as a static summary
3. The audit domain of each PK agent that then provides that information

**Comment by tegefaulkes** [comment]
There's an `EventTarget` leak somewhere. 

```
(node:3176) MaxListenersExceededWarning: Possible EventTarget memory leak detected. 11 abort listeners added to [AbortSignal]. Use events.setMaxListeners() to increase limit
(node:3176) MaxListenersExceededWarning: Possible EventTarget memory leak detected. 11 abort listeners added to [AbortSignal]. Use events.setMaxListeners() to increase limit
(node:3176) MaxListenersExceededWarning: Possible EventTarget memory leak detected. 11 abort listeners added to [AbortSignal]. Use events.setMaxListeners() to increase limit
(node:3176) MaxListenersExceededWarning: Possible EventTarget memory leak detected. 11 abort listeners added to [AbortSignal]. Use events.setMaxListeners() to increase limit
(node:3176) MaxListenersExceededWarning: Possible EventTarget memory leak detected. 11 abort listeners added to [AbortSignal]. Use events.setMaxListeners() to increase limit
```

**Comment by CMCDragonkai** [comment]
I forgot that event targets have limits. We should probably set it to a higher number anyway in js-events because the default limit is quite low for what we are doing. But you think there's a leak? You might want to log out verbosely where all the registrations are, can monkey patch the registration function.

**Comment by tegefaulkes** [comment]
For abort signals we chain them, so the amount of handlers is usually shallow. Even if we don't chain them we usually don't keep adding handlers to the signal. SO Yeah, I think I need to look into this.

**Comment by CMCDragonkai** [comment]
@tegefaulkes has closed off alot of the nodes related issues. We should do another test between Austin and Sydney again to confirm.

Also this issue https://github.com/MatrixAI/Polykey-CLI/issues/47 was related to the same ipv4/ipv6 mapping issue. Was this solved too?

If all of this is done, you can help me with https://github.com/MatrixAI/Polykey/issues/365 and any other quality of life problems like you mentioned above.

**Comment by CMCDragonkai** [comment]
Based on the current set of outstanding issues, there is some quality of life things to consider here.

@amydevs 

* https://github.com/MatrixAI/Polykey/issues/237 - reference this in the new PR for audit domain, make sure to tag @okneigres 

@tegefaulkes some quick things to do:

* https://github.com/MatrixAI/js-quic/issues/66
* https://github.com/MatrixAI/Polykey-CLI/issues/17
* https://github.com/MatrixAI/Polykey/issues/518
* Warning on js-db - can move to node 18 as it still works there

**Comment by tegefaulkes** [comment]
Completed two of the issues. #52 is done but I cut out a bunch of changes when rebasing it and I want to make sure if it was needed or not. There are no mention of what these changes were for in the PR or issue and AFAIK the formatting was fixed in a newer RP already.

I can look at the DB change but it's not critical, same for #17. Either of them can be done later.

**Comment by CMCDragonkai** [comment]
By Monday the https://github.com/MatrixAI/Polykey/issues/628 should be providing audit RPC calls that can be called by the dashboard service @okneigres. The CLI commands can be added in concurrently by @amydevs

**Comment by tegefaulkes** [comment]
Just finished the meeting. Here is a general outline of what was discussed.

Brian
1. Working on the Decentralised hole punching
    1. NCM mostly done
    2. NCM now focuses on holding and making the connections
    3. there are 2 styles, direct connection on host and port, or hole punch with signalling
    4. The Connection map now can hold multiple connections per target node.
    5. The connection with the 'lowest' `connectionIdShared` by string cmp is the one selected from the map when using connections
    6. All connections time out using the usual mechanism
2. Possibly blocked on `NodeGraph` structure expansion.
3. Moving on to `NodeManager` to handle higher level connection logic and ICE.

Oleg
3. He has an API for mapping IP to co-ords on the map
4. We could use a basic vector graphic for the world map.
5. Main things that are left are to put it all together and deploy.
6. TODO
   1. Resolving some issues building with nix.
   2. implement some template for the page
   3. All the components exist but just needs to put everything together and deploy it.
7. blockers
   1. No real blockers right now. SO long as building and deployment proceed smoothly.

Amy
1. Audit domain, getting audit events are fully type safe
2. But the RPC client can't really handle types right now unless you wrap the generic callers.
3. Seems that the async generators @ready decorator doesn't properly work? need to look into that.
4.  Right now the getAuditEvents and getAuditEventsLongRunning are separate. I thin from an API perspective we want these to be the same command where getting ascending with   no limit will extend to long-running and future events.
5. metrics right now are implemented by iterating over the records, or in this case, counting between a range. It's not ideal for performance. I think we'd want to calculate running metrics on insertion. Something to look into.
6. TODO
    1. Just need to get RPC types working with the system.
    2. Handle tracking of generators for if audit domain stops so it can throw.
    3. Also enable the long-running generators to just end with a signal but without error.
7. blockers
    1. None at the moment.

Addie
1. making progress, should be able to publish by EOD or tomorrow.
2. Getting started on the map, can start frame-working the front-end with dummy data.
3. TODO
    1. Working on and further fleshing out CSS for website. Done by today or tomorrow.
    2. Working on map visualisation. Should take a few hours
4. Blockers
    1. Blocked by back-end API for the map data. But can move forward using dummy data.

**Comment by CMCDragonkai** [comment]
If @okneigres gets to AWS deployment please help him @tegefaulkes by pairing up. It can be easier since we have more experience doing nix and docker and AWS.

@amydevs switch to using online running algorithms. Chatgpt this for more information for calculating metric. It does not require running over the entire dataset all the time. See things like running/moving average. L can be streamed.

@addievo the dashboard and website is going to be part of our launch party. It's going to be big and dazzling! We want it to be billboard level.

**Comment by CMCDragonkai** [comment]
Monday, we get an update on the situation from https://github.com/MatrixAI/Polykey/issues/599 and https://github.com/MatrixAI/Polykey/pull/618.

I'll be drawing out the plan for the docs.

**Comment by addievo** [comment]
1. Cloning is broken as of 8/12. Was working on 7/12. RPC times out when trying to clone


addie@Addie-MB-Server:~$ polykey -np /home/addie/Node2 vaults clone vault2 vcnjo8bhfqhusvcd89u12getdqqiqbvf7avfed2u71udcgdl660fg --verbose
INFO:polykey.PolykeyClient:Creating PolykeyClient
INFO:polykey.PolykeyClient.Session:Creating Session
INFO:polykey.PolykeyClient.Session:Setting session token path to /home/addie/Node2/token
INFO:polykey.PolykeyClient.Session:Starting Session
INFO:polykey.PolykeyClient.Session:Started Session
INFO:polykey.PolykeyClient.Session:Created Session
INFO:polykey.PolykeyClient:Starting PolykeyClient
INFO:polykey.PolykeyClient.WebSocketClient:Create WebSocketClient to ::1:46397
INFO:polykey.PolykeyClient.WebSocketClient.WebSocketConnection 0:Start WebSocketConnection
INFO:polykey.PolykeyClient.WebSocketClient.WebSocketConnection 0:Started WebSocketConnection
INFO:polykey.PolykeyClient.WebSocketClient:Created WebSocketClient to wss://[::1]:46397
INFO:polykey.PolykeyClient:Started PolykeyClient
INFO:polykey.PolykeyClient:Created PolykeyClient
INFO:polykey.PolykeyClient.WebSocketClient.WebSocketConnection 0.WebSocketStream 0:Start WebSocketStream
INFO:polykey.PolykeyClient.WebSocketClient.WebSocketConnection 0.WebSocketStream 0:Started WebSocketStream
INFO:polykey.PolykeyClient.WebSocketClient.WebSocketConnection 0.WebSocketStream 0:Destroy WebSocketStream
INFO:polykey.PolykeyClient.WebSocketClient.WebSocketConnection 0.WebSocketStream 0:Destroyed WebSocketStream
INFO:polykey.PolykeyClient:Stopping PolykeyClient
INFO:polykey.PolykeyClient.WebSocketClient:Destroy WebSocketClient on wss://[::1]:46397
INFO:polykey.PolykeyClient.WebSocketClient.WebSocketConnection 0:Stop WebSocketConnection
INFO:polykey.PolykeyClient.WebSocketClient:ErrorWebSocketConnectionLocal: WebSocket Connection local error - Locally closed with code 1000
INFO:polykey.PolykeyClient.WebSocketClient.WebSocketConnection 0:ErrorWebSocketConnectionLocal: WebSocket Connection local error - Locally closed with code 1000
INFO:polykey.PolykeyClient.WebSocketClient.WebSocketConnection 0:Stopped WebSocketConnection
INFO:polykey.PolykeyClient.Session:Stopping Session
INFO:polykey.PolykeyClient.Session:Stopped Session
INFO:polykey.PolykeyClient:Stopped PolykeyClient
ErrorRPCTimedOut: RPC has timed out

The node has adequate permissions to clone 

addie@Addie-MB-Server:~$ polykey -np /home/addie/Node2 vaults scan vcnjo8bhfqhusvcd89u12getdqqiqbvf7avfed2u71udcgdl660fg
vault2  z8WvCNBUwEsrXw9ySPXsuWH pull,clone

**Comment by addievo** [comment]
The "closed during fork" error was observed on NixOS specifically and I am unable to reproduce it now, the cloning error is occuring across platforms on NixOS, Ubuntu and MacOS. I am on commit c8607b9aff37bf2187ce638ac4480ada8999faf6.

**Comment by addievo** [comment]
@tegefaulkes

**Comment by tegefaulkes** [comment]
The problem here is that these commands are failing to make the connection to do their thing. So the connection attempt via `findNode` till usually take longer than the 15 second RPC timeout. 

So the real problem here is that the `RPC` is timing out during long running commands and resulting in a very un-descriptive error. A failure to add a node should take as long as it wants and timeout with a fail to ping or more descriptive error. The same goes for any commands that make a connection.

**Comment by tegefaulkes** [comment]
Here's a basic script I used for the demonstration.

#### starting an agent

1. start an agent with `./polykey agent start`. Enter the password if its the first time starting
2. we can verify its running with `./polykey agent status`


#### creating a vault with secrets

1. create a vault with `./polykey vaults create testvault`.
2. we can list the vaults with `./polykey vaults list`.
2. create a secret we want to add to the vault with `echo somesecret > secret`
3. add it to a vault with `./polykey secrets create ./secret testvault:secret`
4. we can see the secret with `./polykey secrets list testvault`
5. we can see the secret with `./polykey secrets get testvault:secret`
6. lets modify the secret with `echo newsecret > secret`
7. and see it with `./polykey secrets get testvault:secret`.

#### viewing the network

1. Polykey connects to the testnet at startup. we can see the active connections with `./polykey nodes connections`
2. We can list all seen nodes with `./polykey nodes getall`
3. I can form connections to nodes with `./polykey nodes ping <NODEID>` This will find the node within the wider network as well as looking for it in the local network if multicast is enabled.

#### trusting a peer

1. For trust we need a 2nd node to to trust. If running a 2nd node on the same machine we can start it with `./polykey agent start -np ./nodeb`
2. We can see it's nodeId wth `./polykey agent status -np ./nodeb`
3. get the first nodes nodeID with `./polykey agent status`.
4. verify you can connect to this node from nodeA with `./polykey nodes ping <NODEID>`
5. have the nodes discover each other with `./polykey identities discover <NodeIdB>`
5. have the nodes discover each other with `./polykey identities discover -np ./nodeb <NodeIdA>`
5. nodeB needs to trust nodeA to allow it to share vaults. `./polykey identities trust -np ./nodeb <NodeIdA>`
6. and vice versa `./polykey idenitites trust <NodeIdB>`
6. We can now share the nodeA's vault with nodeB with `./polykey vaults share testvault <NodeIdB>`
7. nodeB can now clone the vault with `./polykey vaults clone -np ./nodeb testvault <NodeIdA>`
8. We can now list the vault with `./polykey vaults list -np ./nodeb`
9. and list the secrets `./polykey secrets list testvault -np ./nodeb`
10. and get the secret `./polykey secrets get testvault:secret -np ./nodeb`

**Comment by CMCDragonkai** [comment]
We cannot abide by the default https://nodejs.org/api/events.html#nodeeventtargetsetmaxlistenersn. We may have legitimate reasons for having lots of event listeners on our event targets.

However we also need to keep track of this count internally. This can be part of our diagnostics domain to avoid internal memory and resource leaks by maintaining counters of all the resources being created.

We may be able to make use of some async monitors for this. It needs to be fast and may be runtime dependent. One thing I'd like to start with is tracking floating promises.

**Comment by CMCDragonkai** [comment]
> We cannot abide by the default https://nodejs.org/api/events.html#nodeeventtargetsetmaxlistenersn. We may have legitimate reasons for having lots of event listeners on our event targets.
> 
> However we also need to keep track of this count internally. This can be part of our diagnostics domain to avoid internal memory and resource leaks by maintaining counters of all the resources being created.
> 
> We may be able to make use of some async monitors for this. It needs to be fast and may be runtime dependent. One thing I'd like to start with is tracking floating promises.

A quick solution right now is to raise the number marginally more. What would it need to be to avoid the warnings as of the current state. Are there situations where the deepest callstack may provide us the largest possible number of event listeners?

Also I'm interpreting this as due to our js-contexts usage right? There's no other places where we are dynamically getting lots of event listeners? And this could be a problem if during a large recursion into the same contextified function?

**Comment by tegefaulkes** [comment]
> > We cannot abide by the default https://nodejs.org/api/events.html#nodeeventtargetsetmaxlistenersn. We may have legitimate reasons for having lots of event listeners on our event targets.
> > However we also need to keep track of this count internally. This can be part of our diagnostics domain to avoid internal memory and resource leaks by maintaining counters of all the resources being created.
> > We may be able to make use of some async monitors for this. It needs to be fast and may be runtime dependent. One thing I'd like to start with is tracking floating promises.
> 
> A quick solution right now is to raise the number marginally more. What would it need to be to avoid the warnings as of the current state. Are there situations where the deepest callstack may provide us the largest possible number of event listeners?
> 
> Also I'm interpreting this as due to our js-contexts usage right? There's no other places where we are dynamically getting lots of event listeners? And this could be a problem if during a large recursion into the same contextified function?

I have a new issue that ties into this problem at https://github.com/MatrixAI/js-rpc/issues.

**Comment by tegefaulkes** [comment]
Going through the open issues I've picked out the following that still seem relevant here.

1. Polykey-CLI
    1. MatrixAI/Polykey-CLI/issues/31
    1. MatrixAI/Polykey-CLI/issues/78
    3. MatrixAI/Polykey-CLI/issues/74
    4. MatrixAI/Polykey-CLI/issues/73
    5. MatrixAI/Polykey-CLI/issues/37
    6. General UX testing and polish
    7. MatrixAI/Polykey-CLI/issues/76
    8. MatrixAI/Polykey-CLI/issues/38
    9. MatrixAI/Polykey-CLI/issues/30
2. Polykey
    1. MatrixAI/Polykey/issues/642
    2. MatrixAI/Polykey/issues/653
    3. MatrixAI/Polykey/issues/482
    4. MatrixAI/Polykey-CLI#99
    5. MatrixAI/Polykey-CLI#80
    6. MatrixAI/Polykey/issues/242
    7. Do mdns queries happen passively as well? Can we make a synchronious query of already seen services for our target NodeID when doing `findByMDNS`? It would solve some timeout nuance. Something to discuss
    8. MatrixAI/Polykey/issues/652
    10. MatrixAI/Polykey/issues/651
    11. MatrixAI/MatrixAI-Graph#42
    12. MatrixAI/Polykey/issues/648
    13. MatrixAI/Polykey/issues/641
    14. MatrixAI/Polykey/issues/625
    15. MatrixAI/Polykey/issues/594
    16. MatrixAI/Polykey/issues/564
    17. MatrixAI/Polykey/issues/537
    18. MatrixAI/Polykey/issues/523
    19. MatrixAI/Polykey/issues/476
    20. MatrixAI/Polykey/issues/461
3. js-quic
    1. MatrixAI/js-quic/issues/78
    2. MatrixAI/js-quic/issues/77
    3. MatrixAI/js-quic/issues/83
    4. MatrixAI/js-quic/issues/64

**Comment by CMCDragonkai** [comment]
And this https://github.com/MatrixAI/Polykey-Docs/pull/33.

@amydevs to review this tomorrow.

**Comment by CMCDragonkai** [comment]
Current priorities:

* https://github.com/MatrixAI/Polykey/issues/599 - @amydevs 
* https://github.com/MatrixAI/Polykey/issues/71 - @tegefaulkes

**Comment by tegefaulkes** [comment]
Things are mostly working with the docker integration tests now except for 1 problem.

One of the tests is failing due to a CLI call being made to an agent failing. It's failing with...

```
{"level":"ERROR","keys":"polykey.PolykeyClient.WebSocketClient","msg":"ErrorWebSocketConnectionLocal: WebSocket Connection local error - WebSocket could not open due to internal error"}
{"level":"ERROR","keys":"polykey.PolykeyClient.WebSocketClient.WebSocketConnection 0","msg":"ErrorWebSocketConnectionLocal: WebSocket Connection local error - WebSocket could not open due to internal error"}
{"type":"ErrorWebSocketConnectionLocal","data":{"message":"WebSocket could not open due to internal error","timestamp":"2023-12-20T03:55:11.025Z","data":{"errorCode":1011,"reason":"WebSocket could not open due to internal error"},"cause":{"errno":-111,"code":"ECONNREFUSED","syscall":"connect","address":"127.0.0.1","port":36569},"stack":"ErrorWebSocketConnectionLocal: WebSocket could not open due to internal error\n    at WebSocket.openErrorHandler (/builds/MatrixAI/open-source/Polykey-CLI/node_modules/@matrixai/ws/src/WebSocketConnection.ts:693:16)\n    at Object.onceWrapper (node:events:629:26)\n    at WebSocket.emit (node:events:514:28)\n    at emitErrorAndClose (/builds/MatrixAI/open-source/Polykey-CLI/node_modules/ws/lib/websocket.js:1016:13)\n    at ClientRequest.<anonymous> (/builds/MatrixAI/open-source/Polykey-CLI/node_modules/ws/lib/websocket.js:864:5)\n    at ClientRequest.emit (node:events:514:28)\n    at TLSSocket.socketErrorListener (node:_http_client:501:9)\n    at TLSSocket.emit (node:events:514:28)\n    at emitErrorNT (node:internal/streams/destroy:151:8)\n    at emitErrorCloseNT (node:internal/streams/destroy:116:3)\n    at processTicksAndRejections (node:internal/process/task_queues:82:21)"}}
```

Specifically the `ECONNREFUSED` when making the connection. The agent is running fine here, it works locally so I can only assume it's some weirdness with the DIND networking setup in the CI job.

I'll do a little more digging but this isn't the priority right now. So I may have to shortcut the tests for now. I can see that the connections are being made and the agent isn't crashing so that'll have to be enough for now.

**Comment by tegefaulkes** [comment]
I'm adding https://github.com/MatrixAI/Polykey-CLI/issues/90 to this to track all the changes needed for the CI integration testing

**Comment by tegefaulkes** [comment]
In no particular order, just picking out things to do.

1. MatrixAI/Polykey/issues/642
2. MatrixAI/Polykey/issues/364
3. explore vaults domain optimizations
4. MatrixAI/Polykey-CLI/issues/78
5. MatrixAI/Polykey-CLI/issues/76
6. MatrixAI/Polykey-CLI/issues/73
7. MatrixAI/Polykey-CLI/issues/37

**Comment by tegefaulkes** [comment]
I need to make an issue about tracking down the potential `EventTarget`/`AbortSignal` handler leak. Right now we get warnings about it but I don't think it's a major issue.

**Comment by tegefaulkes** [comment]
So I've been looking into why the vaults and adding secrets is slow. From what I can tell the problem stems from two parts.

1. Each time we write a file we do a commit. I don't know how involved read/write wise making a commit with isomorphic git is but it's likely doing a few operations. From the vaults perspective, most of the time is spent on committing.
2. Just writing and reading a file in EFS is about 30 ops/s That is already not that fast. Even just opening an file and getting a fd is about 100 ops/s

I think we need to add some more bench-marking in EFS to narrow down the performance bottlenecks and try to optimise it.

**Comment by tegefaulkes** [comment]
New issue for the `EventTarget` leak at https://github.com/MatrixAI/js-quic/issues/80.

I've found the problem, just needs some discussion.

**Comment by CMCDragonkai** [comment]
I noticed that the `polykey` ECR image is still just `polykey` and it's name hasn't changed to `polykey-cli`. Like in:

```
registry_image='015248367786.dkr.ecr.ap-southeast-2.amazonaws.com/polykey'
```

@amydevs it should be changed to `polykey-cli` for our ECR images since `polykey` is going to be reserved for the library stuff.

This would align with the release output names given that we produce `polykey-cli` in:

![image](https://github.com/MatrixAI/Polykey-CLI/assets/640797/3efc5800-9f89-48a2-a799-c54819eb395b)

I've also removed the image publishing documentation from `README.md` because it's no longer relevant now, it's being moved to `Polykey-Docs` under the development guide.

@brynblack you might want to keep awareness about this.

**Comment by CMCDragonkai** [comment]
Also I've regenerated the docs, it hadn't been done since this repo was first created. We would probably want to generate the docs automatically in our CI/CD instead of leaving it here in our source to avoid manual creation. It also reveals useful information like broken links too.

**Comment by CMCDragonkai** [comment]
Newly deployed docs looks alot better:

![image](https://github.com/MatrixAI/Polykey-CLI/assets/640797/6a1fa2ff-93b9-4290-8a7e-5fbefd571ce8)

@brynblack feel free to give feedback as to whether auto generated docs is useful at all. I don't really read it myself since I just go through the source code.

**Comment by CMCDragonkai** [comment]
@tegefaulkes can you add the relevant issues into this epic via zenhub? And then update whether they are in-progress or otherwise. Keep this epic organised.

**Comment by CMCDragonkai** [comment]
The conclusion of this issue should update the version of `polykey-cli` to `0.X`. So right now we are on `0.1.Y`. So it should be whatever the latest `0.1.Y` we end up with.

I reckon the next version after that will be `1.0.0`.

**Comment by tegefaulkes** [comment]
> @tegefaulkes can you add the relevant issues into this epic via zenhub? And then update whether they are in-progress or otherwise. Keep this epic organised.

I've been doing this already. I've been adding the new ones I've created and updated the status for the issues I've been working on. I haven't touched the status for others however.

**Comment by CMCDragonkai** [comment]
I've noticed that the `integration:prerelease` doesn't run normally. And therefore when we make a release tag, an image is pushed to the GitLab container registry, but it doesn't update the `testnet` tag. Which is confusing because the `mainnet` tag ends up being "later" then the `testnet` tag.

![image](https://github.com/MatrixAI/Polykey-CLI/assets/640797/3f5dcf30-0f92-46ba-9495-14a2332a33b3)

https://gitlab.com/MatrixAI/open-source/Polykey-CLI/container_registry

In my infrastructure planning, I'm making diagram about how our distribution infrastructure works.

**Comment by CMCDragonkai** [comment]
Some feedback on the error reporting concept: https://github.com/MatrixAI/js-rpc/issues/57#issuecomment-1913884490.

I think we should swap the cause chain.

**Comment by tegefaulkes** [comment]
Going over the videos and discord comments I've compiled a shortlist of things to address.

1. ~Running Polykey-CLI with just the help text has a bit of a delay. It's slow to start, but I don't think slow in general. I think that the PKG or commander Has some startup cost involved. Probably PKG since it needs to unpack some stuff to even run.~ https://github.com/MatrixAI/Polykey-CLI/issues/102
2. ~When Oauthed on github the app is refered to as `Polykey Desktop`. it should be either just `Polykey` or `Polykey CLI` for the CLI specifically.~ https://github.com/MatrixAI/Polykey-CLI/issues/103
3. ~Sync node graph failed. Are the seed nodes running? Check AWS, otherwise I need to test locally. If that works then I need more details.~ https://github.com/MatrixAI/js-quic/commits/staging/
4. ~Allow for untrusted notifications but as a separate list from the trusted ones.~ https://github.com/MatrixAI/Polykey-CLI/issues/105
5. ~Identities search has no separation between listed identities. The formatting needs to be fixed.~ https://github.com/MatrixAI/Polykey-CLI/issues/106
6. ~When getting RPC errors we're meant to unwrap the remoteRPC error and just show the cause. This links into the verbosity affecting the error message.~ https://github.com/MatrixAI/Polykey-CLI/issues/107
7. ~trusting identities is case sensitive, we may want to fuzzy search and or recommend a similar identity if the one provided is wrong somehow.~ https://github.com/MatrixAI/Polykey-CLI/issues/108
8. ~Should have a command to reset local state~ https://github.com/MatrixAI/Polykey-CLI/issues/109
9. ~Should have the ability to list out all identity providers available~ https://github.com/MatrixAI/Polykey-CLI/issues/110
10. ~The identity claim still times too quickly~ Reopened https://github.com/MatrixAI/Polykey/issues/588
11. ~update to node 20 and switch out `vercel/pkg` for https://github.com/yao-pkg/pkg~ https://github.com/MatrixAI/Polykey-CLI/issues/111

I'll be creating new issues or adding to existing issues to address these.

**Comment by tegefaulkes** [comment]
We get the following warning when running the CLI

```
(node:610526) [DEP0112] DeprecationWarning: Socket.prototype._handle is deprecated
(Use `node --trace-deprecation ...` to show where the warning was created)
```

@amydevs mentioned it's due to MDNS. I think we need to look into removing the warning. I don't see any issues about this however so maybe we should make a new one and look into it?

It seems that MDNS is getting the handle (fd?) of the socket to modify it. I can see us needing direct access to the socket in native code for `js-quic` as well for a performance upgrade. So solving this can be applied there as well.

**Comment by tegefaulkes** [comment]
even shorter list: 

1. https://github.com/MatrixAI/Polykey/issues/588
2. https://github.com/MatrixAI/Polykey/issues/453
3. https://github.com/MatrixAI/Polykey-CLI/issues/76
4. https://github.com/MatrixAI/Polykey/issues/537
5. https://github.com/MatrixAI/Polykey-CLI/issues/111
6. https://github.com/MatrixAI/Polykey-CLI/issues/22
7. https://github.com/MatrixAI/Polykey-CLI/issues/6

low low priority

1. https://github.com/MatrixAI/Polykey/issues/648
2. https://github.com/MatrixAI/Polykey/issues/653
3. https://github.com/MatrixAI/Polykey/issues/461
4. https://github.com/MatrixAI/Polykey-CLI/issues/17

@amy pick from these and address them

**Comment by tegefaulkes** [comment]
`Secrets env` command is completed now. I'll be moving on to working on the short short list OR fixing and re-enabling windows and mac CI test jobs

**Comment by amydevs** [comment]
pr MatrixAI/Polykey-Docs#56 should be ready now

**Comment by tegefaulkes** [comment]
I've noticed that the test `NodeManager › with peers in network › findNode by signalled connections › handles offline nodes` fails randomly and more often than I'd like. We need to look into it.

**Comment by tegefaulkes** [comment]
We're getting close to completing the main stuff. @amydevs is looking into the issues with discovery we found while testing. I'm thinking I should look into making `polykey` more tolerant of network failures and changes. This can tie into https://github.com/MatrixAI/Polykey/issues/461 as well.

I've already made some fixes to `js-quic` to handle network problems but it's far from comprehensive. `Polykey` itself needs to handle the `QUICServer` going down for network reasons. It should even support the network being unavailable for extended periods.

This ties in nicely with https://github.com/MatrixAI/Polykey/issues/461 since we'd need to support not having networking running for extended periods of time, but also support dynamically starting or stopping the `QUICServer` whenever we want. Either in respose to the user requesting it or having to restart it when the network changes or comes back up. Also, being able to toggle the `QUICServer` at will can be seen as a kind of stealth mode.

I just did some testing with `Polykey-CLI`. It seems things are tolerant of this already so the changes I made before seem more than fine for now. I tested this by starting a node and just turning off wifi. I did the expected and timeout after the timeout time. Re-enabling the wifi and pinging a seed node caused it to re-connect to the Polykey network just fine as expect. The process did not crash at any point while doing this. Previous crashing due to network issues may have been fixed when I fixed a race condition relating to this just recently.

That said, there may be corner cases where the socket could just fail for one reason or another. The question is how tolerant do we want to be about this? How does this handle switching between wifi networks or interfaces? If we have two active interfaces and one starts to fail then do we seamlessly switch, or are we *just* tolerant of that? I suppose all theses questions will bear out with more rigorous user testing.

So based on that quick test I think as things are we're fine and tolerant of network failures in that the process will no crash in these cases. I don't think there's anything to address about that right now but we should keep it in the back of our mind with further testing.

**Comment by tegefaulkes** [comment]
Priorities going forward..

### Discovery problems

Right now there are problems with discovery
1. Bug when processing duplicate and missing cryptolinks
2. No feedback when doing discovery, back-grounded or active discovery.
3. Currently periodic re-discovery happening. So discovery always needs to be manually triggered.
4. GG doesn't seem to be updated when processing new information during discovery.
5. We want to seamlessly trigger discovery on any friends/followers on Github 
6. No removal of old or invalid nodes from the gestalt.

Reviewing the repos I found the following old issues that relate to these problems.

- https://github.com/MatrixAI/Polykey/issues/691 Newly created by Amy. Outlines periodic rediscovery of vertices and automatic discovery of peers. Points 5, 3.
- https://github.com/MatrixAI/Polykey/issues/462 Outlines a few features, fixes and optimisations that need to be addressed for discovery. Including periodic rediscovery. 3. 
- https://github.com/MatrixAI/Polykey/issues/328 Outlines rediscovery mostly and better error handling for discovery. Points 1, 3.
- https://github.com/MatrixAI/Polykey-CLI/issues/163 Bug reporting points 6, 1, 2, 4.
- https://github.com/MatrixAI/Polykey-CLI/issues/162 Addresses point 2.

Looks like all points are addressed across this. Some double up. Should we create a single PR addressing all of these? Seems like a decent re-work/update of discovery. All of these are on `Polykey` except for point 2 which is both front-end and back-end.

### Performance problems.

There are 2 performance issues. 

1. Slow startup time of commands. This is likely to be two things. Either Commander doing a bunch of runtime processing for constructing commands. Or PKG doing a bunch of unpacking to run. https://github.com/MatrixAI/Polykey-CLI/issues/102
2. The EFS runs pretty slow which results in very slow vaults performance. https://github.com/MatrixAI/js-encryptedfs/issues/76

### Lingering issues

We still need to finish up the following active issues
- https://github.com/MatrixAI/Polykey-CLI/issues/22
- Small problem with recent change in MDNS causing test failures in Polykey.

### Moving forward

Moving forward we want to focus on any UI/UX issues and bugs we encounter during user testing.

**Comment by tegefaulkes** [comment]
I'm going to make a new epic for tracking discovery problems.

**Comment by CMCDragonkai** [comment]
@pablo.padillo this is the main important engineering issue atm.

Status update:

1. Remaining subissues here are not part of CLI beta 30, they can be moved out to be worked on in the backlog.
2. The main bug discovered during your testing with externals and internals in relation to having double claims should be fixed. Pending that this is all part of a released PK CLI executable - if it is not, it requires a 0.3.1 release.
3. If so, then you must re-attempt another demo run - but this time without bothering with removing your old claims. It's fine to have multiple claims to the same identity from multiple nodes.

```
  I1
 /  \
N1   N2

Even if N1 is destroyed, from the perspective of I1, and N2, the entire gestalt is all 3 vertexes.

Therefore if N3 and I2 exists.

I2
|
N3

And I1...I2 were connected in some way.

Then N3 should be able to discover I2 and I1, and also discover N1 and N2.

Thus N3 should be able to connect to N2 - either via manual discovery or automatic discovery.
```

The work that @brian.botha and @amydevs is on right now, are all UX oriented features that help automate or reduce the amount of work for the entire demo process.

This includes:

1. Being able for N3 to share with ANY vertex of I1, so it can share to I1 or N2 or N1, and it achieves the same thing. So that N1 and N2 can both pull/clone the vault. This reduces the demo process steps. - [ENG-132](https://linear.app/matrix-ai/issue/ENG-132/vault-sharing-should-be-possible-with-a-gestalt-identity-id)
2. Background discovery should be automatic between I2 gestalt and I1 gestaslt. So that it should not be needed for N3 to manual-discover of N2/I1. Background discovery should be interpretable and predictable to the first-order - the immediate neighbourhood. - ISSUE?
3. Being able to have notifications be completely delay tolerant. - [ENG-37](https://linear.app/matrix-ai/issue/ENG-37/failure-to-send-notification-for-vaults-share-should-fail-in-a-less) [https://github.com/MatrixAI/Polykey/pull/695](https://github.com/MatrixAI/Polykey/pull/695)

These 3 things may come after version 0.3.1.

**Comment by CMCDragonkai** [comment]
So upon release of 0.3.1, remove those other remaining subissues, as I consider the beta release to be done. @pablo.padillo should be redoing the demo with the understanding the more than 1 claim is fine to work with. Subsequent release of PK should be addressing the 3 items posted above.

**Comment by tegefaulkes** [comment]
> @pablo.padillo this is the main important engineering issue atm.
> 
> Status update:
> 
> 1. Remaining subissues here are not part of CLI beta 30, they can be moved out to be worked on in the backlog.
> 2. The main bug discovered during your testing with externals and internals in relation to having double claims should be fixed. Pending that this is all part of a released PK CLI executable - if it is not, it requires a 0.3.1 release.
> 3. If so, then you must re-attempt another demo run - but this time without bothering with removing your old claims. It's fine to have multiple claims to the same identity from multiple nodes.
> 
> ```
>   I1
>  /  \
> N1   N2
> 
> Even if N1 is destroyed, from the perspective of I1, and N2, the entire gestalt is all 3 vertexes.
> 
> Therefore if N3 and I2 exists.
> 
> I2
> |
> N3
> 
> And I1...I2 were connected in some way.
> 
> Then N3 should be able to discover I2 and I1, and also discover N1 and N2.
> 
> Thus N3 should be able to connect to N2 - either via manual discovery or automatic discovery.
> ```
> 
> The work that @brian.botha and @amydevs is on right now, are all UX oriented features that help automate or reduce the amount of work for the entire demo process.
> 
> This includes:
> 
> 1. Being able for N3 to share with ANY vertex of I1, so it can share to I1 or N2 or N1, and it achieves the same thing. So that N1 and N2 can both pull/clone the vault. This reduces the demo process steps. - [ENG-132](https://linear.app/matrix-ai/issue/ENG-132/vault-sharing-should-be-possible-with-a-gestalt-identity-id)
> 2. Background discovery should be automatic between I2 gestalt and I1 gestaslt. So that it should not be needed for N3 to manual-discover of N2/I1. Background discovery should be interpretable and predictable to the first-order - the immediate neighbourhood. - ISSUE?
> 3. Being able to have notifications be completely delay tolerant. - [ENG-37](https://linear.app/matrix-ai/issue/ENG-37/failure-to-send-notification-for-vaults-share-should-fail-in-a-less) [Backgrounding of Notifications Domain Polykey#695](https://github.com/MatrixAI/Polykey/pull/695)
> 
> These 3 things may come after version 0.3.1.

https://linear.app/matrix-ai/issue/ENG-298/follow-permission-and-social-links-during-discovery has been created to track point 2. above.

**Comment by CMCDragonkai** [comment]
@CryptoTotalWar I think upon closing this issue, we should be planning for a hard launch. You should create a new issue for that and associate subissues for a hard launch. I'm not sure if it should go into Polykey - but we could associate issues to the MatrixAI-Graph for any general work.

**Comment by CryptoTotalWar** [comment]
NOTES Rough Draft WIP. Will clean up. 



Critical Issues 

1. https://github.com/MatrixAI/Polykey/issues/592 Pretty major problem, results in nodes randomly crashing. "General Fixes for connection stability" - important from a network stability standpoint. We don't want the program or network to fail on you. 

2. https://github.com/MatrixAI/Polykey/issues/597 - An oddity and doesn't really break things, but should be looked into. "was producing multiple node certificates when it should have only been producing one" - Minotr bug dont include 

3. https://github.com/MatrixAI/js-rpc/pull/47 https://github.com/MatrixAI/Polykey/issues/588 https://github.com/MatrixAI/Polykey/pull/589 - Critical from a usability standpoint. Otherwise not strictly broken. "increasing the Authentication Timeout window for PK Identities Authenticate - important from a usability standpoint 

4. https://github.com/MatrixAI/Polykey-CLI/issues/44 - Critical from a usability stand point. "General PK CLI polish and fixes" - pretty major (I will say this is a feature enhancement from a usability standpoint). 

5. https://github.com/MatrixAI/js-timer/issues/15 - macro task resource leak. "just a fix where the timers are not being cleaned up properly... a resource leak where they weree lingering in memory when they should have been done cleaned up" - major - resource leaks always a problem and the memory would pile up and crash the node so an improtant fix. 

Needed but functional without 

- major fix where we fixed NAT hole punching in multiple directions whereas in certain network situations there could be single direction hole punching whereas it should have been symmetric in certain situation. - Important 

https://github.com/MatrixAI/Polykey/issues/605 - Not super big issue, Right now it means more stricter nats can't be punched to. I have an idea for a fix. Not strictly a major error, just something we did not want to be broken. Important for reliability of usage in different network situations. 

- There was a fix where we weren't taking into account an OLD node ID when authenticating a connection. So when a node renewed its ID (renewed its key) you should still be able to authenticate a connection with an old node but we werent handling that properly. https://github.com/MatrixAI/Polykey/issues/593 --- VERY MINOR 

- "small fixes to the timeout middleware" https://github.com/MatrixAI/js-rpc/issues/19 https://github.com/MatrixAI/js-rpc/pull/42  - NOt important 

A lot of this was just fixing up bugs. 
 
 
 "Major Feature enhancement 
 
 - we overhauled how we discovered nodes in the network and made connection between each other with nat hole punching in a more decentralized way whereas before it was more centralized where all hole punching went through seed nodes but in this case any node could be used to whole punch and discover other nodes https://github.com/MatrixAI/Polykey/pull/618
 
 
 - we did add the Polykey ENV command which allows injecting secrets into the environment when running a command," which simulate env files in a more secure encrypted manner. 
 https://github.com/MatrixAI/Polykey-CLI/issues/31
 
 - Amy did a lot of work related to standardizing the output formats for the CLI. https://github.com/MatrixAI/Polykey-CLI/issues/22 
 
 Features WIP 
 
 - Audit domain - https://github.com/MatrixAI/Polykey-CLI/issues/177
 
 - Feedback to discovery to inspect the discovery q but also discovery events as they happen when triggering discount https://github.com/MatrixAI/Polykey-CLI/issues/162 - discovery feedback

**Comment by tegefaulkes** [comment]
0.3.1 has been release a few days ago. I'm going to resolve this issue now.

**Comment by CryptoTotalWar** [comment]
### Engineering Updates Summary for Blog Post

#### Critical Fixes
- **Connection Stability**: Implemented general fixes to enhance network stability, addressing critical issues where nodes could crash unexpectedly. This fix is crucial for ensuring reliability. [Issue #592](https://github.com/MatrixAI/Polykey/issues/592) - **Important**
  
- **Authentication Enhancements**: Increased the Authentication Timeout window for PK Identities Authenticate to improve user experience during the login process. [PR MatrixAI/Polykey-Docs#47](https://github.com/MatrixAI/js-rpc/pull/47), [Issue #588](https://github.com/MatrixAI/Polykey/issues/588), [PR #589](https://github.com/MatrixAI/Polykey/pull/589) - **Important**

- **CLI Usability**: Enhancements and polish applied to the Polykey CLI to improve usability. [Issue MatrixAI/Polykey-Docs#44](https://github.com/MatrixAI/Polykey-CLI/issues/44) - **Important**

- **Resource Leak Fix**: Addressed a significant issue where timers were not being properly cleaned, leading to potential memory buildup and crashes. [Issue MatrixAI/Polykey-Docs#15](https://github.com/MatrixAI/js-timer/issues/15) - **Important**

#### Feature Enhancements
- **NAT Hole Punching**: Refined our NAT hole punching mechanisms to support more symmetric network configurations, crucial for the reliability of node communications. [Issue #605](https://github.com/MatrixAI/Polykey/issues/605) - **Important**

- **Node Discovery Overhaul**: Decentralized the node discovery process to enhance network connectivity and reduce reliance on seed nodes. [PR #618](https://github.com/MatrixAI/Polykey/pull/618) - **Major Feature Enhancement**

- **Environment Variable Command**: Added the Polykey ENV command for secure injection of secrets into the environment, simulating a more secure method of handling environment variables. [Issue MatrixAI/Polykey-Docs#31](https://github.com/MatrixAI/Polykey-CLI/issues/31) - **Major Feature Enhancement**

- **CLI Output Standardization**: Standardized the output formats of the CLI, improving interface consistency across commands. [Issue MatrixAI/Polykey-Docs#22](https://github.com/MatrixAI/Polykey-CLI/issues/22) - **Important**

#### Features in Progress (WIP)
- **Audit Domain**: Developing an audit domain for better monitoring and tracking of network activities. [Issue #177](https://github.com/MatrixAI/Polykey-CLI/issues/177) - **WIP**

- **Discovery Feedback**: Enhancing feedback mechanisms for node discovery to allow real-time insights into the discovery process. [Issue #162](https://github.com/MatrixAI/Polykey-CLI/issues/162) - **WIP**

#### Not Included
- Minor bugs and enhancements not critical to the user experience were deemed not important for the blog post, ensuring focus on impactful updates.

This summary will guide the blog post update regarding the CLI enhancements and fixes since the beta release. Please review and confirm the inclusion of these points. - @tegefaulkes

**Comment by CMCDragonkai** [comment]
I think that should go to your blog issue? This issue is closed. @CryptoTotalWar
